### PR TITLE
[S3] Use AWS Core SDK credentials for auth

### PR DESCRIPTION
**Pull request by tkroman**
Use AWS Core SDK credentials mechanism to fully support all possible auth variants.

This will automatically fix e.g. #367 and solve all the potential problems with token refresh etc.

I took a liberty to change the config as well. Probably will need more docs as well.

Just let me know if this is even the right approach. Tests didn't break (after some adjustments), but compatibility (source/binary) is certainly broken.

I also ran `scalafmt` on the repo root, hence some seemingly unrelated changes.

### CODE REVIEW

**Comment by tkroman** [comment]
Here's the ammonite script which I ran on an EC2 instance to check that everything is fine: https://gist.github.com/tkroman/b41a54ff0753fe8be13474e2829eb688. As you can see, no credentials are used, but this works perfectly.

**Comment by shyamsalimkumar** [comment]
Just curious, but is it possible to implement the same but avoid having to include `aws-java-sdk` as a dependency or is that too much of a hassle? There might've been a good reason for removing the SDK dependency in the first place.

**Comment by tkroman** [comment]
Too much. With AWS SDK “we” support all the available ways to get credentials - from .aws/credentials, properties, env variables, ec2 metadata service, anonymous access & profile credentials. Some of these support expiration, others do not etc. This would take too much effort to implement and support.

**Comment by avietrov** [comment]
Would it make sense to split into this PR into two? as you've mentioned, scalafmt did lots of changes.

**Comment by avietrov** [comment]
I see you've used AWSCredentials. Just for my understanding, for whatever reason only AWSSessionCredentials work in my case, using simple credentials fails with signing errors. Is this something specific to my setup?

**Comment by tkroman** [comment]
I’m not sure if PRs that don’t pass formatting checks (if there is such a thing) can be accepted.

On account of your credentials question - sorry, need more context.

**Comment by tkroman** [comment]
Oh wait, I got the question. Sure, there is a difference. Session credentials also include a token, which is ignored by basic credentials, but required in the process of signing. Basically, this is what I’m talking about - too much to understand, too much to implement anew, too much to take care about.

**Comment by idzivinskyi** [comment]
@tkroman it looks so good. When do you plan to release it?

**Comment by tkroman** [comment]
@idzivinskyi when the akka team has time to review, I think. 

+1 on a PR could make a difference. I'm going to go and shamelessly self-promote on gitter :D.

**Comment by tkroman** [comment]
@johanandren @raboof (tagging you because looks like you are the active maintainers, sorry if no) any chance of getting some feedback on this?

File: s3/src/main/scala/akka/stream/alpakka/s3/scaladsl/S3Client.scala [hunk 1]
```diff
@@ -48,10 +45,13 @@ object S3Client {
   def apply()(implicit system: ActorSystem, mat: Materializer): S3Client =
     new S3Client(S3Settings(system.settings.config))
 
-  def apply(credentials: AWSCredentials, region: String)(implicit system: ActorSystem, mat: Materializer): S3Client = {
-    val settings = S3Settings
-      .apply(system.settings.config)
-      .copy(awsCredentials = credentials, s3Region = region)
+  def apply(credentials: AWSCredentialsProvider, region: String)(implicit system: ActorSystem,
```
Comment (johanandren) [inline s3/src/main/scala/akka/stream/alpakka/s3/scaladsl/S3Client.scala]: Could we keep the old signature as well, and just call this method wrapping the AWSCredentials with a static provider, that way we avoid breaking everyones code.

**Comment by johanandren** [comment]
Sorry about the long delay for review, we have been very busy with other things.

File: s3/src/main/scala/akka/stream/alpakka/s3/scaladsl/S3Client.scala (hunk 1 above)
Comment (tkroman) [reply s3/src/main/scala/akka/stream/alpakka/s3/scaladsl/S3Client.scala]: Will do. However, the code relying on config will still break, do you want me to do a compatibility layer for that?

**Comment by tkroman** [comment]
Merged, rebased on the latest master, ran scalafmt. Should be 👌 now.

**Comment by tkroman** [comment]
Also, added compat layer for config, so old users should not suffer from this change (I hope) Check out s3/src/test/scala/akka/stream/alpakka/s3/S3SettingsSpec.scala & comments in reference.conf.

**Comment by idzivinskyi** [comment]
Hi @tkroman @johanandren, 

I have a problem when I use this approach. It looks like after expiration of token or after a big timeout between requests ```s3Client.download``` starts to return new token instead of the original file. (I have one instance of client per application)

What should I do in that case? Do I need to init s3Client for each request or I need to add an additional configuration to aws auth?

Thanks.

**Comment by tkroman** [comment]
1) wdym “token instead of file”, is that a typo or a really really weird behavior?
2) what is your current configuration?

**Comment by idzivinskyi** [comment]
```
val s3Client: S3Client = S3Client()
```

configuration from config:
```
stream.alpakka.s3 {
    buffer = "memory"
    aws {
      credentials.provider = default
      default-region = "us-east-2"
    }
  }
```

**Comment by tkroman** [comment]
I don't think tokens should expire when used properly. Can you ping me on gitter so we can discuss this in more detail? I'm not affiliated with the Akka team, just feeling sort of responsible for the behavior my changes might be related to.

**Comment by idzivinskyi** [comment]
I will link a solution here: https://github.com/akka/alpakka/pull/545/files

**Pull request merged**